internal/graph/rate_limited_check_resolver.go

```diff
@@ -0,0 +1,176 @@
+	}, []string{"grpc_service", "grpc_method", "queue_name"})
+)
+
+func NewRateLimitedCheckResolver(
+	config RateLimitedCheckResolverConfig) *RateLimitedCheckResolver {
+	ticker := time.NewTicker(config.TimerTickerFrequency)
+	rateLimitedCheckResolver := &RateLimitedCheckResolver{
+		config:           config,
+		ticker:           ticker,
+		lowPriorityQueue: make(chan bool),
+		medPriorityQueue: make(chan bool),
+		done:             make(chan bool),
+	}
+	rateLimitedCheckResolver.setMediumConfig(config)
+	rateLimitedCheckResolver.delegate = rateLimitedCheckResolver
+	go rateLimitedCheckResolver.runTicker()
+	return rateLimitedCheckResolver
+}
+
+func (r *RateLimitedCheckResolver) setMediumConfig(config RateLimitedCheckResolverConfig) {
+	mediumRange := config.LowPriorityLevel - config.MediumPriorityLevel
+	r.mediumPriorityLevel2 = config.MediumPriorityLevel + mediumRange/8
+	r.mediumPriorityLevel3 = config.MediumPriorityLevel + mediumRange/4
+	r.mediumPriorityLevel4 = config.MediumPriorityLevel + mediumRange/2
+	r.mediumPriorityShaperFreq2 = config.MediumPriorityShaper / 2
+	r.mediumPriorityShaperFreq3 = config.MediumPriorityShaper / 4
+}
+
+func (r *RateLimitedCheckResolver) SetDelegate(delegate CheckResolver) {
+	r.delegate = delegate
+}
+
+func (r *RateLimitedCheckResolver) Close() {
+	r.done <- true
+}
+
+func (r *RateLimitedCheckResolver) nonBlockingSend(signalChan chan bool) {
+	select {
+	case signalChan <- true:
+		// message sent
+	default:
+		// message dropped
+	}
+}
+
+func (r *RateLimitedCheckResolver) handleTimeTick(lowPriorityQueueCounter uint32) uint32 {
+	lowPriorityQueueCounter++
+	r.nonBlockingSend(r.medPriorityQueue)
+	if lowPriorityQueueCounter >= r.config.LowPriorityShaper {
+		lowPriorityQueueCounter = 0
+		r.nonBlockingSend(r.lowPriorityQueue)
+	}
+	return lowPriorityQueueCounter
+}
+
+func (r *RateLimitedCheckResolver) runTicker() {
+	lowPriorityQueueCounter := uint32(0)
+	for {
+		select {
+		case <-r.done:
+			r.ticker.Stop()
+			close(r.done)
+			close(r.medPriorityQueue)
+			close(r.lowPriorityQueue)
+			return
+		case <-r.ticker.C:
+			lowPriorityQueueCounter = r.handleTimeTick(lowPriorityQueueCounter)
+		}
+	}
+}
+
+func (r *RateLimitedCheckResolver) shouldWait(currentNumDispatch uint32) bool {
+	delta := currentNumDispatch - r.config.MediumPriorityLevel
+	if currentNumDispatch < r.mediumPriorityLevel2 {
+		return delta%(r.config.MediumPriorityShaper) == 0
+	}
+	if currentNumDispatch < r.mediumPriorityLevel3 {
+		return delta%(r.mediumPriorityShaperFreq2) == 0
+	}
+	if currentNumDispatch < r.mediumPriorityLevel4 {
+		return delta%(r.mediumPriorityShaperFreq3) == 0
+	}
+	return true
+}
+
+func (r *RateLimitedCheckResolver) ResolveCheck(ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+	start := time.Now()
+	queueName := "high_priority"
+
+	currentNumDispatch := req.DispatchCounter.Add(1)
+	if currentNumDispatch > r.config.MediumPriorityLevel {
```

Comment:
We currently cannot find the root request (the first check) when resolving a check. If we are able to individually identify root request, we can put them at the back of the queue each time they cross a threshold. This way we wouldn't need multiple queues.

I might be missing context here on why above suggestion wouldn't work but would love to understand it.